admin/test/test_release.py

```diff
@@ -661,3 +669,704 @@ def test_publish_non_release_fails(self):
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_metadata_file = os.path.join(self.target_key, 'repodata',
+                                              'filelists.xml.gz')
+
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+            existing_metadata_file: 'old_metadata_content',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=self.repo_contents),
+            packages=self.packages,
+            flocker_version=self.dev_version,
+            distro_name=self.operating_systems[0]['distro'],
+            distro_version=self.operating_systems[0]['version'],
+        )
+
+        # This tests that filelists.xml.gz, a metadata file which is not the
+        # index, still has the content 'old_metadata_content', even though
+        # CreateRepo changes this content. This shows that this file which
+        # already existed on S3 is not re-uploaded.
+        self.assertEqual(
+            aws.s3_buckets[self.target_bucket][existing_metadata_file],
+            existing_s3_keys[existing_metadata_file])
+
+    def test_new_metadata_files_uploaded(self):
+        """
+        Calling :func:`update_repo` uploads new repository metadata files to
+        S3.
+        """
+        index_path = os.path.join(self.target_key, 'repodata', 'repomd.xml')
+        existing_s3_keys = {
+            index_path: 'old_metadata_index',
+        }
+
+        aws = FakeAWS(
+            routing_rules={},
+            s3_buckets={
+                self.target_bucket: existing_s3_keys.copy(),
+            },
+        )
+
+        self.update_repo(
+            aws=aws,
+            yum=FakeYum(),
+            rpm_directory=self.rpm_directory,
+            target_bucket=self.target_bucket,
+            target_key=self.target_key,
+            source_repo=self.create_fake_repository(files=self.repo_contents),
+            packages=self.packages,
+            flocker_version=self.dev_version,
+            distro_name=self.operating_systems[0]['distro'],
+            distro_version=self.operating_systems[0]['version'],
+        )
+
+        repodata_files = [
+            key for key in aws.s3_buckets[self.target_bucket] if
+            key.startswith(os.path.join(self.target_key, 'repodata'))]
+
+        # What matters is that there is more than just the index.
+        # Other tests cover the creation and modification of particular
+        # metadata filenames, but these change because of `createrepo` options
+        # such as whether to use hashes in filenames or not.
```

Comment:
We are using a fake CreateRepository, so we know the names of all the metadata files that are created.